File: EduLink.Domain/Entities/States/ReservadoState.cs
using EduLink.Domain.Enums;

namespace EduLink.Domain.Entities.States;

public class ReservadoState : SlotState
{
    public override EstadoSlot Nombre => EstadoSlot.Reservado;

    public override void Cancelar(SlotHorario slot)
    {
        slot.CupoActual--;
        slot.EstadoInterno = new DisponibleState();
    }

    // Anda lleno
}

File: app/domain/entities/states/slot/reservado_state.rb
require_relative './slot_state'
require_relative '../../../enums/estado_slot'
require_relative './disponible_state'

module Entities
  module States
    module Slot
      class ReservadoState < SlotState
        def nombre
          Enums::EstadoSlot::RESERVADO
        end

        def cancelar(slot)
          # Si cancelan y estaba lleno, se libera un cupo y vuelve a estar disponible.
          slot.cupo_actual -= 1
          slot.estado_interno = DisponibleState.new
        end

        # Anda lleno, así que no se puede reservar. No hacemos nada.
      end
    end
  end
end
